<?xml version="1.0" ?>
<launch>
  <!-- Launch file for the admittance controller -->

  <!-- TOPICS FROM THE LOWER-LEVE CONTROLLER -->

  <!-- Communication with the arm, reading the state of the arm and commanding velocities -->
  <arg name="TOPIC_ARM_COMMAND" default="/ur5/ur5_cartesian_velocity_controller_sim/command_cart_vel" />
  <arg name="TOPIC_ARM_POSE"  default="/ur5/ur5_cartesian_velocity_controller_sim/ee_pose"  />
  <arg name="TOPIC_ARM_TWIST" default="/ur5/ur5_cartesian_velocity_controller_sim/ee_twist" />


  <!-- TOPICS AND SETTINGS FOR ADMITTANCE CONTROLLER -->

  <!-- Communication with the Force/Torque sensor, reading the force/torque -->
  <!-- The wrench is specified in robotiq_force_torque_frame_id -->
  <arg name="TOPIC_EXTERNAL_WRENCH"  default="/admittance_control/wrench" />
  <arg name="WRENCH_FILTER_FACTOR"   default="0.1" />
  <arg name="FORCE_DEAD_ZONE_THRES"  default="5.0" />
  <arg name="TORQUE_DEAD_ZONE_THRES" default="0.5" />

  <!-- An input topic that can be used for control (expected to be in ur5_arm_base_link)-->
  <arg name="TOPIC_CONTROL_WRENCH" default="/admittance_control/wrench_control" />
  
  <!-- An input topic to receive desired velocity from DS only the linear motion-->
  <arg name="TOPIC_DS_VELOCITY" default="/ds_desired_velocity" />


  <!-- The received wrench by the FT-sensor and the controller will be published for dubugging/visualization -->
  <!-- They will be specified in ur5_arm_base_link -->
  <arg name="TOPIC_EXTERNAL_WRENCH_ARM_FRAME" default="/admittance_control/transformed/external_wrench" />
  <arg name="TOPIC_CONTROL_WRENCH_ARM_FRAME"  default="/admittance_control/transformed/control_wrench"  />

  <!-- admittance ratio is a value between zero and one that modulated the input force as to control the stiffness-->
  <arg name="TOPIC_ADMITTANCE_RATIO"  default="/admittance_control/admittance_ratio"  />

  <!-- Loading the admittance parameters from a yaml file -->
  <arg name="ADMITTANCE_PARAMS_FILE" default="$(find admittance_control)/config/admittance_params_real.yaml" />
 
  <group ns="admittance_control">
    <!-- Converting args into params -->
    <param name="topic_arm_pose"                   value="$(arg TOPIC_ARM_POSE)"                 />
    <param name="topic_arm_twist"                  value="$(arg TOPIC_ARM_TWIST)"                />
    <param name="topic_arm_command"                value="$(arg TOPIC_ARM_COMMAND)"              />
    <param name="topic_external_wrench"            value="$(arg TOPIC_EXTERNAL_WRENCH)"          />
    <param name="topic_control_wrench"             value="$(arg TOPIC_CONTROL_WRENCH)"           />
    <param name="topic_ds_velocity"                value="$(arg TOPIC_DS_VELOCITY)"              />
    <param name="topic_external_wrench_arm_frame"  value="$(arg TOPIC_EXTERNAL_WRENCH_ARM_FRAME)"/>
    <param name="topic_control_external_arm_frame" value="$(arg TOPIC_CONTROL_WRENCH_ARM_FRAME)" />
    <param name="topic_admittance_ratio"           value="$(arg TOPIC_ADMITTANCE_RATIO)"          />

    <rosparam file="$(arg ADMITTANCE_PARAMS_FILE)" command="load"/>

    <node name="admittance_controller" pkg="admittance_control" 
    type="admittance_controller_node" output="screen"/>

    </group>

</launch>
